src/matrix-utils.ts

```diff
@@ -365,39 +358,58 @@ export async function createRoom(
  * Returns an absolute URL to that will load Element Call with the given room
  * @param roomId ID of the room
  * @param roomName Name of the room
- * @param password e2e key for the room
+ * @param encryptionSystem what encryption (or EncryptionSystem.Unencrypted) the room uses
  */
 export function getAbsoluteRoomUrl(
   roomId: string,
+  encryptionSystem: EncryptionSystem,
   roomName?: string,
-  password?: string,
+  viaServers?: string[],
 ): string {
   return `${window.location.protocol}//${
     window.location.host
-  }${getRelativeRoomUrl(roomId, roomName, password)}`;
+  }${getRelativeRoomUrl(roomId, encryptionSystem, roomName, viaServers)}`;
 }
 
 /**
  * Returns a relative URL to that will load Element Call with the given room
  * @param roomId ID of the room
  * @param roomName Name of the room
- * @param password e2e key for the room
+ * @param encryptionSystem what encryption (or EncryptionSystem.Unencrypted) the room uses
  */
 export function getRelativeRoomUrl(
   roomId: string,
+  encryptionSystem: EncryptionSystem,
   roomName?: string,
-  password?: string,
+  viaServers?: string[],
 ): string {
   // The password shouldn't need URL encoding here (we generate URL-safe ones) but encode
   // it in case it came from another client that generated a non url-safe one
-  const encodedPassword = password ? encodeURIComponent(password) : undefined;
-  if (password && encodedPassword !== password) {
-    logger.info("Encoded call password used non URL-safe chars: buggy client?");
-  }
+  const encryptionPart = ((): string => {
+    switch (encryptionSystem?.kind) {
+      case E2eeType.SHARED_KEY: {
+        const encodedPassword = encodeURIComponent(encryptionSystem.secret);
+        if (encodedPassword !== encryptionSystem.secret) {
+          logger.info(
+            "Encoded call password used non URL-safe chars: buggy client?",
+          );
+        }
+        return "&" + PASSWORD_STRING + encodedPassword;
+      }
+      case E2eeType.PER_PARTICIPANT:
+        return "&" + PER_PARTICIPANT_STRING + "true";
+      case E2eeType.NONE:
+        return "";
+    }
+  })();
 
+  const roomIdPart = `roomId=${roomId}`;
+  const viaServersPart = viaServers
+    ? viaServers.map((s) => "&" + VIA_SERVERS_STRING + s).join("")
+    : "";
   return `/room/#${
     roomName ? "/" + roomAliasLocalpartFromRoomName(roomName) : ""
-  }?roomId=${roomId}${password ? "&" + PASSWORD_STRING + encodedPassword : ""}`;
+  }?${roomIdPart}${encryptionPart}${viaServersPart}`;
```

Comment:
I don't like that we're expanding the PASSWORD_STRING hack to now also include these PER_PARTICIPANT_STRING and VIA_SERVERS_STRING constants. Maybe this is a sign that we should finally implement a proper function `UrlParams => string` or `UrlParams => URLSearchParams` or something similar, that would allow us to consolidate all information about the URL params into UrlParams.ts?